codex/stores/sqlitestore.nim

```diff
@@ -0,0 +1,204 @@
+
+  if getCachedRes.get.isSome:
+    trace "Retrieved block from cache", cid
+    return getCachedRes
+
+  let
+    blkKey = blockKey(cid)
+    getRes = await self.datastore.get(blkKey)
+
+  if getRes.isErr:
+    trace "Unable to read block", key = blkKey.id, error = getRes.error.msg
+    return failure "Unable to read block"
+
+  let
+    dataOpt = getRes.get
+
+  if dataOpt.isNone:
+    return success Block.none
+
+  without blk =? Block.new(cid, dataOpt.get), error:
+    trace "Unable to construct block from data", cid, error = error.msg
+    return failure error
+
+  let
+    putCachedRes = await self.cache.putBlock(blk)
+
+  if putCachedRes.isErr:
+    trace "Unable to store block in cache", cid, error = putCachedRes.error.msg
+
+  return success blk.some
+
+method putBlock*(
+  self: SQLiteStore,
+  blk: Block): Future[?!void] {.async.} =
+  ## Write a block's contents to the database with key based on blk.cid,
+  ## save a copy to the cache
+  ##
+
+  if blk.isEmpty:
+    trace "Empty block, ignoring"
+    return success()
+
+  let
+    cid = blk.cid
+    blkKey = blockKey(cid)
+    putRes = await self.datastore.put(blkKey, blk.data)
+
+  if putRes.isErr:
+    trace "Unable to store block", key = blkKey.id, error = putRes.error.msg
+    return failure "Unable to store block"
+
+  let
+    putCachedRes = await self.cache.putBlock(blk)
+
+  if putCachedRes.isErr:
+    trace "Unable to store block in cache", cid, error = putCachedRes.error.msg
+
+  return success()
+
+method delBlock*(
+  self: SQLiteStore,
+  cid: Cid): Future[?!void] {.async.} =
+  ## Delete a block from the database and cache
+  ##
+
+  trace "Deleting block from database", cid
+  if cid.isEmpty:
+    trace "Empty block, ignoring"
+    return success()
+
+  let
+    blkKey = blockKey(cid)
+    delRes = await self.datastore.delete(blkKey)
+
+  if delRes.isErr:
+    trace "Unable to delete block", key = blkKey.id, error = delRes.error.msg
+    return failure "Unable to delete block"
+
+  let
+    delCachedRes = await self.cache.delBlock(cid)
+
+  if delCachedRes.isErr:
+    trace "Unable to delete block from cache", cid,
+      error = delCachedRes.error.msg
+
+  return success()
+
+method hasBlock*(
+  self: SQLiteStore,
+  cid: Cid): Future[?!bool] {.async.} =
+  ## Check if a block exists in the database
+  ##
+
+  trace "Checking database for block existence", cid
+  if cid.isEmpty:
+    trace "Empty block, ignoring"
+    return true.success
+
+  return await self.datastore.contains(blockKey(cid))
+
+method listBlocks*(
+  self: SQLiteStore,
+  onBlock: OnBlock): Future[?!void] {.async.} =
+  ## Process list of all blocks in the database via callback.
+  ## This is an intensive operation
+  ##
+
+  for kd in self.datastore.query(allBlocks):
+    let
+      (key, _) = await kd
+      cidRes = Cid.init(key.name)
+
+    if cidRes.isOk:
+      let
+        cid = cidRes.get
+
+      try:
+        await onBlock(cid)
```

Comment:
`type OnBlock` is marked with `upraises: []`

https://github.com/status-im/nim-codex/blob/7cfb157d327624d848a09c2df4eda20ea677a8e4/codex/stores/blockstore.nim#L24

So do we really need to `try/except` here?

I copied that logic from `FSSstore`, but locally confirmed that it compiles and runs correctly when `await onBlock(cid)` is not wrapped in `try`.

Reply:
I think we should be fine removing the try/except from both

Reply:
Done.